find xor paths; grid input to model

YijieYin opened the pull request:
This PR introduces:
- a function that finds 'XOR' computation from 3-layered paths; 
- functions that provide grid-like input to the model and visualise the output; 

### XOR 
Based on [Wang et al. 2024](https://www.biorxiv.org/content/10.1101/2024.09.24.614724v2), a three-layer network with two inputs can do XOR when there are two exciters forwarding the inputs separately, in addition to one inhibitor which receives from both of the inputs. 

#### How 
I start with the two exciters and get their upstream input. I get the input only to `exciter1`, and input only to `exciter2` into separate pools (for separate forwarding). Then I loop through the inhibitors, to see whose input includes neurons in both pools. In the end, I check if these exciters and inhibitor have the same downstream target. 

### grid-like input 
This is inspired by the [TensorFlow playground](https://playground.tensorflow.org/). It's to vary the strength of two inputs in a grid-like manner, to observe how the output changes.  
I'd like to use this to examine the role of non-linearity in (interpreting) biological neural circuits.

The functions would be used as follows: 
```
grid_inputs, grid_coords = make_grid_inputs(v1, v2, num_layers)
batch_out = ml_model(grid_inputs)
plot_output_grid(batch_out, grid_coords, selected_indices)
```
which then gives a heatmap (like the one in the TensorFlow playground) for each timestep. 

### Testing
- Unit tests added. 
- Ran `python -m unittest discover tests`, and all tests passed successfully.

### Notes 
- I'm not very sure about the best way to store the outputs for XOR: currently it has its own class. 
  - advantage: the roles of the neurons are clear (input, exciter, inhibitor, output), 
  - drawback: if the original paths dataframe has weights, this information isn't preserved. 
 
Alternatively, I could return a list of paths dataframes. Paths dataframes are like the output of [this function](https://connectome-interpreter.readthedocs.io/en/latest/modules/path_finding.html#connectome_interpreter.path_finding.find_path_iteratively): it's a dataframe with columns pre, post, weight, layer. 
- Please don't pay too much attention to the tests. They were written by Claude.

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1087,164 @@ def path_for_ngl(path):
     return pd.concat(dfs)
+
+
+def make_grid_inputs(v1, v2, num_layers: int, grid_size: int = 10, timepoints: int | list = 0, device=None) -> Tuple[torch.Tensor, List[Tuple[float, float]]]:
```
"Modern" numpy recommends using `numpy.typing.NDArray` rather than `np.ndarray` ([doc](https://numpy.org/doc/stable/reference/typing.html)):
```python
import numpy.typing as npt

def fun(xs: npt.NDArray, ...): ...
```

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1087,164 @@ def path_for_ngl(path):
+    """
+
+    # first check if length of v1 and v2 are the same
+    if len(v1) != len(v2):
```
do you need them to be 1-dimensional too?

YijieYin replied:
Good point, modified.

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1087,164 @@ def path_for_ngl(path):
+    """
+    selected_index = to_nparray(selected_index)
+
+    def plot_heatmap(index):
+        # Create the plot
+        plt.figure(figsize=figsize)
+        im = plt.imshow(heatmaps[index-1], cmap='viridis', origin='lower',
+                        extent=[0, 1, 0, 1])
+        plt.colorbar(im, label='Output')
+        plt.xlabel('v1 Activation')
+        plt.ylabel('v2 Activation')
+        plt.title(f'Output Grid for {selected_index}')
+        plt.show()
```
I tend to prefer using the object-oriented model
```suggestion
    def plot_heatmap(index: int) -> plt.Figure:
        # Create the plot
        fig,  ax = plt.subplots(1, 1, figsize=figsize)
        im = aximshow(heatmaps[index-1], cmap='viridis', origin='lower',
                        extent=[0, 1, 0, 1])
        ax.colorbar(im, label='Output')
        ax.xlabel('v1 Activation')
        ax.ylabel('v2 Activation')
        ax.title(f'Output Grid for {selected_index}')
        fig.show()  # see below... maybe you don't want it here
        return fig  # see below
```

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1087,164 @@ def path_for_ngl(path):
+                               description='Timepoint', continuous_update=True)
+
+    # Link the slider to the plotting function
+    display(widgets.interactive(plot_heatmap, index=slider))
```
How many heatmaps do you have and how long does it take to render each image?

You could get a more responsive UI if you first create all the `plt.Figure` objects first (using a factory function, see above) and then this `display` is just responsible for retrieving the correct figure. Something like this:
```python
n = len(heatmaps)
# pre-generate all the figures, but don't display them
figs = [plot_heatmap(idx) for idx in range(1, n + 1)]  # see above
slider = widgets.IntSlider(...)
display(widgets.interactive(
    lambda index: figs[index].show(),  # show the chosen figure
    index=slider
))
```

jftsang replied:
Fair enough – I think it might depend also on what Jupyter you're using (vanilla, or Colab, or whatever VS Code gives, or ...).

jftsang commented on tests/test_utils.py:
```diff
@@ -57,3 +60,165 @@ def test_handle_zero_colsums(self):
+        circuits = find_xor(df)
+        max_possible = len(list(itertools.combinations(middles, 3)))
+        self.assertLessEqual(len(circuits), max_possible)
+
+
+# Add these test cases to your test suite
+if __name__ == "__main__":
+    pytest.main([__file__])
```
```suggestion
```

jftsang reviewed (changes requested):
Some superficial comments, haven't looked at the XOR stuff yet

YijieYin commented:
Hi @jftsang, thanks a lot for reviewing. I've addressed your comments. Please feel free to have another look :)

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -354,7 +355,7 @@ def modify_coo_matrix(sparse_matrix, input_idx=None, output_idx=None, value=None
 
 
-def to_nparray(input_data: int | float | list | set | np.ndarray | pd.Series, unique: bool = True) -> np.ndarray:
+def to_nparray(input_data: int | float | list | set | npt.NDArray | pd.Series, unique: bool = True) -> npt.NDArray:
```
Just an idea, take it or leave it:
```python
from numbers import Real  # or `Number`, if you want complex numbers too
from typing import Collection  # finitely-sized iterables, not necessarily ordered (e.g. `set`)

# types that can be made into a numeric numpy array
arrayable = Real | Collection[Real]

def to_nparray(
    input_data: arrayable, unique: bool = True
) -> npt.NDArray: ...
```

([doc for Real](https://docs.python.org/3/library/numbers.html#numbers.Real), [doc for Collection](https://docs.python.org/3/library/collections.abc.html#collections.abc.Collection))

If you don't want to use the ABCs you can still assign a type alias:
```python
arrayable = int | float | list | set | npt.NDArray | pd.Series
```

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+    return inputs_tensor, grid_coords
+
+
+def plot_output_grid(grid_outputs: torch.Tensor | npt.NDArray, grid_coords: List[Tuple[float, float]], selected_index: int | List[int], figsize: tuple = (10, 8),
+                     cmap: str = 'viridis', xlab: str = 'v1 Activation', ylab: str = 'v2 Activation', title: str = None):
```
Nit:
```suggestion
def plot_output_grid(
    grid_outputs: torch.Tensor | npt.NDArray,
    grid_coords: List[Tuple[float, float]],
    selected_index: int | List[int],
    *,  # see note
    figsize: tuple = (10, 8),
    cmap: str = 'viridis', 
    xlab: str = 'v1 Activation', 
    ylab: str = 'v2 Activation', 
    title: str | None = None,
):
```
I prefer this style for arglists: it's longer but it makes the code "invariant under refactoring" – it's easier to add or remove arguments or rename the function without having to re-indent everything.

For functions with large numbers of kwargs, especially plotting functions, I like to [use `*` to force arguments to be given as keywords](https://stackoverflow.com/questions/59661042/what-do-single-star-and-slash-do-as-independent-parameters): again, easier to change things if you spell them out.

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+    # Middle layer nodes
+    exciter1: str  # Takes input from input1 only
+    exciter2: str  # Takes input from input2 only
+    inhibitor: str        # Takes input from both inputs
```
Nit:
```suggestion
    inhibitor: str  # Takes input from both inputs
```

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+            xlab=xlab, ylab=ylab, title=title))
+
+
+@dataclass
```
Does this stuff want to be in a separate module? 

I try to avoid having a `utils.py` module that grows too large; business logic like this probably deserves to be in its own module. I try to reserve `utils` for silly low-level things like string formatting or routine calculations.

YijieYin replied:
Good point, moved.

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+            "The input DataFrame should have exactly 2 unique layers, 1 and 2")
+
+    # check column names of paths
+    if not {'pre', 'post', 'sign', 'layer'}.issubset(paths.columns):
```
Optionally `not s1 <= s2` instead of `not s1.issubset(s2)`. Although it can be confusing since `not s1 <= s2` isn't the same thing as `s2 < s1`!

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+        raise ValueError(
+            "The input DataFrame should have columns ['pre', 'post', 'sign', 'layer']")
+
+    # check values of signs: if it's a subset of {1, -1}
+    if set(paths.sign) > {1, -1}:
```
See above comment. I think you want
```python
if not set(paths.sign) <= {1, -1}:
```

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+            f"The input DataFrame should have values 1 (excitatory) or -1 (inhibitory) in the 'sign' column, but got {set(paths.sign)}")
+
+    # define variables ----
+    circuits = []
```
When initializing an empty list, specify its eventual type:
```suggestion
    circuits: list[XORCircuit] = []
```

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+    # define variables ----
+    circuits = []
+
+    exciters = paths.pre[(paths.layer == 2) & (paths.sign == 1)].unique()
```
Maybe `df["col"]` rather than `df.col`. (`.sign` isn't a method of `DataFrame`, but it sure looks like one!)

https://www.youtube.com/watch?v=no06H03hj9k

But it is a lot to change.

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+    l1 = paths[paths.layer == 1]
+    l2 = paths[paths.layer == 2]
+
+    exciter_us_dict = {exc: set(l1.pre[l1.post == exc]) for exc in exciters}
```
Use type annotations if you can, rather than [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation):
```suggestion
    exciter_us_dict: dict[..., set[...]] = {exc: set(l1.pre[l1.post == exc]) for exc in exciters}
```

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+                         for inh in inhibitors}
+
+    # main algorithm ----
+    for e1, e2 in itertools.combinations(exciters, 2):
```
You never need them the other way round? Or do you want `permutations` instead?

YijieYin replied:
Thanks! I don't want them the other way around actually. The two exciters are essentially symmetric in the function, and in what they are meant to do.

jftsang commented on connectome_interpreter/utils.py:
```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+        onlye1 = exciter_us_dict[e1] - common
+        onlye2 = exciter_us_dict[e2] - common
+
+        for i in inhibitors:
+            e1_i_intersect = onlye1 & inhibitor_us_dict[i]
+            e2_i_intersect = onlye2 & inhibitor_us_dict[i]
+            if (len(e1_i_intersect) != 0) and (len(e2_i_intersect) != 0):
+                targets = exciter_ds_dict[e1] & exciter_ds_dict[e2] & inhibitor_ds_dict[i]
+                if len(targets) > 0:
+                    circuits.append(XORCircuit(input1=list(e1_i_intersect), input2=list(e2_i_intersect),
+                                               exciter1=e1, exciter2=e2, inhibitor=i, output=list(targets)))
```
Use guard clauses (btw PyCharm can do this refactor automatically)
```suggestion
        for i in inhibitors:
            e1_i_intersect = onlye1 & inhibitor_us_dict[i]
            e2_i_intersect = onlye2 & inhibitor_us_dict[i]
            if (len(e1_i_intersect) == 0) or (len(e2_i_intersect) == 0):
                continue
            targets = exciter_ds_dict[e1] & exciter_ds_dict[e2] & inhibitor_ds_dict[i]
            if not targets:
                continue
            circuits.append(
                XORCircuit(
                    input1=list(e1_i_intersect), 
                    input2=list(e2_i_intersect),
                    exciter1=e1, 
                    exciter2=e2, 
                    inhibitor=i, 
                    output=list(targets)
                )
            )
```

jftsang commented on tests/test_utils.py:
```diff
@@ -57,3 +60,160 @@ def test_handle_zero_colsums(self):
             self.example_coo, updates_df=zero_updates_df, re_normalize=True)
         column_sums = np.array(result.sum(axis=0))[0]
         self.assertAlmostEqual(column_sums[1], 0.0)
+
```
not checking this for now

YijieYin commented:
Hi @jftsang, thanks so much - I learned about guard clauses and typing in the content of the function (I thought it was just a thing for arguments), among other things. I think I've addressed all your comments. Merging now.

jftsang commented:
Happy to help!

jftsang commented on connectome_interpreter/activation_maximisation.py:
```diff
@@ -11,6 +12,9 @@
 from .compress_paths import result_summary
 from .utils import adjacency_df_to_el, get_activations, to_nparray
 
+# types that can be made into a numeric numpy array
+arrayable = Real | Collection[Real]
```
You could `from .utils import arrayable` too, for consistency

YijieYin replied:
Good point. Thanks!

jftsang commented on tests/test_utils.py:
```diff
@@ -1,3 +1,4 @@
+import pytest
```
Is this needed?

YijieYin replied:
No 🙈 was just scared that I'd break something if I removed it 🙈. Removed now.
